### ISSUE TITLE
GM_download doesn't download a set of files if provided in a short time period

### ISSUE THREAD

**Issue by double-beep** [Tampermonkey/tampermonkey#1113]
Asked [on Stack Overflow](https://stackoverflow.com/q/64044097), where [another user mentioned](https://stackoverflow.com/questions/64044097/userscript-gm-download-not-downloading-all-files#comment113254749_64044097) that they can't reproduce the bug in another userscript manager.

Trying to download a number of files in a short time period results in many files not being downloaded at all. For me, only half of the files are downloaded.

## Specifications

- Chromium: 87.0.664.66
- TM: 4.11.6120
- OS: Windows 10.0.18363.1256

## Script (copied from the post above)
```js
// ==UserScript==
// @name          Image Save Test
// @namespace     ist
// @include       *testimages.juliaimages.org/
// @version       1.0.0
// @run-at        document-end
// @noframes      true
// @grant         GM_download
// ==/UserScript==

var currentDownloads = 0;
var queue = [  'https://testimages.juliaimages.org/thumbnails/autumn_leaves.png',
               'https://testimages.juliaimages.org/thumbnails/blobs.png',
               'https://testimages.juliaimages.org/thumbnails/cameraman.png',
               'https://testimages.juliaimages.org/thumbnails/fabio_color_512.png',
               'https://testimages.juliaimages.org/thumbnails/earth_apollo17.png',
               'https://testimages.juliaimages.org/thumbnails/fabio_gray_256.png',
               'https://testimages.juliaimages.org/thumbnails/hela-cells.png',
               'https://testimages.juliaimages.org/thumbnails/lake_gray.png',
               'https://testimages.juliaimages.org/thumbnails/house.png',
               'https://testimages.juliaimages.org/thumbnails/jetplane.png',
               'https://testimages.juliaimages.org/thumbnails/lake_color.png',
               'https://testimages.juliaimages.org/thumbnails/lena_gray_16bit.png',
               'https://testimages.juliaimages.org/thumbnails/lighthouse.png',
               'https://testimages.juliaimages.org/thumbnails/mandril_color.png',
               'https://testimages.juliaimages.org/thumbnails/mandril_gray.png',
               'https://testimages.juliaimages.org/thumbnails/mountainstream.png',
               'https://testimages.juliaimages.org/thumbnails/peppers_color.png',
               'https://testimages.juliaimages.org/thumbnails/moonsurface.png',
               'https://testimages.juliaimages.org/thumbnails/peppers_gray.png',
               'https://testimages.juliaimages.org/thumbnails/toucan.png',
               'https://testimages.juliaimages.org/thumbnails/pirate.png'];

(async function() {
   while (queue.length > 0) {
      if (currentDownloads > 5) {
         await sleep(200);
         continue;
      }
      
      var item = queue.shift();
      
      (function(_item) {
         GM_download({
            url: _item,
            name: _item.match(/([0-9A-Za-z _-]+)(?=\.png)/)[0],
            saveAs: false,
            onerror: function(error) {
               queue.unshift(_item);
               currentDownloads--;
            },
            onload: function() {
               currentDownloads--;
            }
         });
         
         currentDownloads++;
      })(item);
   }
})();

function sleep(ms) {
   return new Promise(resolve => setTimeout(resolve, ms));
}
```

**Comment by derjanb** [comment]
Note: a workaround is to set "Download Mode" to "Browser API".

**Comment by AlttiRi** [comment]
It's a **browser specific behaviour** due to you try to save multiple files at the same moment. You need to add a delay between file saving.

You can use `GM_xmlhttpRequest` for concurrent downloading, but saving of the blobs should be one by one with 100+ ms delay.

_For me_ with that user script:
TM downloads 15 of 21 files.
VM downloads 19 of 21 files.

It only says that TM works faster.

Theoretically that userscript _may_ _(or not)_ download all files, especially if the connection speed is slow.

Such code works normally (but not concurrently).
```js
const urls = [...];

(async function() {
    let delay;
    for (const url of urls) {
        const name = url.match(/[^\/]+$/)[0];
        delay = sleep(100);
        await download(url, name);
        await delay;
    }
})();

function download(url, name) {
   return new Promise((resolve, reject) => {
       GM_download({
            url,
            name,
            saveAs: false,
            onerror: reject,
            onload: resolve
         });
   });
}

function sleep(ms) {
   return new Promise(resolve => setTimeout(resolve, ms));
}
```

**Comment by AlttiRi** [comment]
> but saving of the blobs should be one by one with 100+ ms delay within the same browser's tab.

You can run multiple scripts with 100 ms saving delay in different tabs, even in different windows (tabs are throttled), and all files will be saved. But if you decrease the delay down to 50 ms and run the script within one tab you will face the files losing.
So, you can _safely_ save up to 10 files if you use 1 tab/window in a second, but you can save up to 30 files in a second if you use 3 tabs/windows. 

@derjanb you should note this moment if you gonna fix this "bug". **_If it is a bug._**

UPD: I tested it with the manual downloading (`fetch` + `<a download href="blob:...">`) from a content script (not with TM or VM). So if the file saving (the click on `<a download href="blob:...">`) in `GM_download` happens in the background script, probably, you can save only up to 10 files in a second independently of the count of the tabs with running the script in it.

But anyway it should be applied for userscripts that use `GM_xmlhttpRequest` (not `GM_download`) with the manual blob downloading:
```js
function downloadBlob(blob, name) {
    const anchor = document.createElement("a");
    anchor.setAttribute("download", name || "");
    anchor.href = URL.createObjectURL(blob);
    anchor.click();
}
```

**Comment by AlttiRi** [comment]
Here is **the correct implementation** of this. With the semaphore.

The limited parallel fetching, the sequential saving.

<details>
  <summary>Screens</summary>

![1](https://user-images.githubusercontent.com/16310547/103612593-d926ed00-4f35-11eb-8dd0-30bce37b6298.png)

![2](https://user-images.githubusercontent.com/16310547/103612601-dcba7400-4f35-11eb-9ff8-30f8379236ee.png)

![3](https://user-images.githubusercontent.com/16310547/103612612-de843780-4f35-11eb-8e9c-6aaf66987fdb.png)      
</details>

```js
// ==UserScript==
// @name        Semaphored Bulk Downloads
// @namespace   GitHub
// @match       https://example.com/*
// @connect     testimages.juliaimages.org
// @grant       GM.xmlHttpRequest
// ==/UserScript==

const urls = [ 'https://testimages.juliaimages.org/thumbnails/autumn_leaves.png',
               'https://testimages.juliaimages.org/thumbnails/blobs.png',
               'https://testimages.juliaimages.org/thumbnails/cameraman.png',
               'https://testimages.juliaimages.org/thumbnails/fabio_color_512.png',
               'https://testimages.juliaimages.org/thumbnails/earth_apollo17.png',
               'https://testimages.juliaimages.org/thumbnails/fabio_gray_256.png',
               'https://testimages.juliaimages.org/thumbnails/hela-cells.png',
               'https://testimages.juliaimages.org/thumbnails/lake_gray.png',
               'https://testimages.juliaimages.org/thumbnails/house.png',
               'https://testimages.juliaimages.org/thumbnails/jetplane.png',
               'https://testimages.juliaimages.org/thumbnails/lake_color.png',
               'https://testimages.juliaimages.org/thumbnails/lena_gray_16bit.png',
               'https://testimages.juliaimages.org/thumbnails/lighthouse.png',
               'https://testimages.juliaimages.org/thumbnails/mandril_color.png',
               'https://testimages.juliaimages.org/thumbnails/mandril_gray.png',
               'https://testimages.juliaimages.org/thumbnails/mountainstream.png',
               'https://testimages.juliaimages.org/thumbnails/peppers_color.png',
               'https://testimages.juliaimages.org/thumbnails/moonsurface.png',
               'https://testimages.juliaimages.org/thumbnails/peppers_gray.png',
               'https://testimages.juliaimages.org/thumbnails/toucan.png',
               'https://testimages.juliaimages.org/thumbnails/pirate.png'];

class Semaphore {
    constructor(max = 1, delay = 0) {
        if (max < 1) {
            max = 1;
        }
        this.max = max;
        this.delay = delay;
        this.count = 0;
        this.queue = [];
    }

    acquire() {
        let promise;
        if (this.count < this.max) {
            promise = Promise.resolve();
        } else {
            promise = new Promise(resolve => {
                this.queue.push(resolve);
            });
        }
        this.count++;
        return promise;
    }

    release() {
        sleep(this.delay)
            .then(_ => {
                if (this.queue.length > 0) {
                    const resolve = this.queue.shift();
                    resolve();
                }
                this.count--;
            });
    }
}


const semaphore = new Semaphore(5);
const mutex = new Semaphore(1, 100);

document.body.onclick = main;
main();

async function main() {
  for (const url of urls) {
    handleUrl(url);
  }
}

async function handleUrl(url) {
    const blob = await fetchBlobSynchronized(url);
    const name = url.match(/[^\/]+$/)[0];
    downloadBlobSynchronized(blob, name);
}

async function fetchBlobSynchronized(url) {
    await semaphore.acquire();
    const blob = await fetchBlob(url);
    semaphore.release();
    return blob;
}

async function downloadBlobSynchronized(blob, name) {
  await mutex.acquire();
  downloadBlob(blob, name);
  mutex.release();
  console.log("Downloaded", name);
}


async function fetchBlob(url) {
    const response = await new Promise((resolve, reject) => {
        GM.xmlHttpRequest({
            url,
            responseType: "blob",
            onload: resolve,
            onerror: reject,
        });
    });
    const {response: blob} = response;
    return blob;
}

function downloadBlob(blob, name) {
    const anchor = document.createElement("a");
    anchor.setAttribute("download", name || "");
    anchor.href = URL.createObjectURL(blob);
    anchor.click();
    setTimeout(_ => URL.revokeObjectURL(blob), 30000);
}

function sleep(ms) {
   return new Promise(resolve => setTimeout(resolve, ms));
}
```

Open https://example.com/. To rerun click at the center.

**Comment by AlttiRi** [comment]
You an also check the thing I talked about.

Open 3 windows with https://example.com/ and reload them fast at one moment. All 63 files will be downloaded successfully.
But if you try to use `const mutex = new Semaphore(1, 100);` with `50` (ms) you will get the file losing even for one tab.

**Comment by AlttiRi** [comment]
It's strange, but today I can't reproduce the bug in VM with the topic starter's script.
The bug only appears then I change `currentDownloads` limit from  `> 5` to `> 14`.

But this script really works much slower in VM than in TM.
It takes **0.5 seconds** in TM, and **12.5 seconds** in VM. (when the images are cached by the browser)

TM downloads blobs in the background script, but VM download them in web page context (It requires to resend the blob two times: background script -> content script -> web page script). Also VM mounts the anchor to DOM before the click.


In comparison my script work fine within ~2.3 seconds in both VM and TM (100 ms interval between the file saving (21 files are total)).

**Comment by AlttiRi** [comment]
BTW, the main reason why VM is so slow is the incorrect sleep time (`await sleep(200);` in the loop). It takes 3 seconds instead of 200 ms.

@tophf, check this (the topic starter's script):

TM (~00.200):
![__Screenshot](https://user-images.githubusercontent.com/16310547/103723706-b7843f00-4fe3-11eb-81b2-fba6831bcab9.png)

VM (~03.000):
![__VM_Screenshot](https://user-images.githubusercontent.com/16310547/103723716-bc48f300-4fe3-11eb-8e1b-99dc9867ba08.png)

**Comment by double-beep** [comment]
@AlttiRi I'm not looking for workarounds/solutions with longer timeouts. I want to know why TM *misses* some files which should have been queued.

**Comment by AlttiRi** [comment]
Khm-kmh.
[VM with the fixed bug](https://github.com/violentmonkey/violentmonkey/commit/ceae9a79c16699dc99d13527dee2ae2ca1b6fc04) I talked about [above](https://github.com/Tampermonkey/tampermonkey/issues/1113#issuecomment-755044544), downloads 17 of 21 images for the first run, and only **10** of 21 for _every_ other runs (when images are cached).

Both extensions work about the same now.

So I was fully right in my assumptions.
That script worked in VM only because VM has (had) the bug that slows down _(significantly)_ the work speed.

Why — I have already answered. It's browser specific behaviour when you try to save a lot of files at one moment.

Probably this topic should have the label not "bug", but "enhancement".

**Comment by derjanb** [comment]
This issue should be fixed at TM BETA 4.12.6126. 😅 I'm now limiting the number of downloads to 3 per 500ms. This works good here and I think the overall speed is not that important. Especially since setting download mode to "Browser API" fixes the complete issue.

I'm not sure how long it'll take for the Chrome Webstore team to review the new version, but in the meantime, you can download it from [here](https://www.tampermonkey.net/crx/tampermonkey_beta_4_12_6126.crx).

**Comment by AlttiRi** [comment]
The funny moment is that _now_ TM works normally with the script from the first message, but VM does not (it downloads only 10 of 21 files) — checked with the latest VM compiled from the source code.
